### DISCUSSION TITLE
Does isDirty work controlled components?

### DISCUSSION THREAD

**Discussion by rickAllDev** [react-hook-form/react-hook-form#2299] [Q&A] [answered]
I'm using the controller with material-ui switches in a form. isDirty is not catching changes to those form elements. Is this as expected?

**Comment by bluebill1049** [comment, marked as answer]
not expected, do you have a CSB?

**Reply by rickAllDev** [reply]
https://codesandbox.io/s/tender-davinci-ul95o?file=/src/App.js

Actually it does seem to catch some changes but it doesn't seem to be very accurate. If you click the 'cloud update' switch a few times and watch the console, you'll see what I mean

**Reply by bluebill1049** [reply]
make sure you read those value before the render phase, as we have wrapped with proxy to improve performance of re-render.

`console.log(formState.isDirty)`

before render.

**Reply by rickAllDev** [reply]
not sure I understand why that works, but it does. thanks for the response.

**Reply by bluebill1049** [reply]
Take a look at the formState in the doc: https://react-hook-form.com/api#formState there is an **important** section.
